# genomes metadata variables
OTU_name = "OTU"
ncbi_id_name = "NCBI_ID"
fasta_name = "fasta"
genome_id_name = "genome_id"
bp_length_name = "bp_length"
novelty_category_name = "novelty_category"
abundance_balanced_name = "abundance_balanced"
known_strain_name = "known_strain"
index_name = "index"

tax_id_name = "tax_id"
tax_name = "taxname"
tax_level_name = "tax_level"
species_name = "species"
genus_name = "genus"
family_name = "family"
kingdom_name = "kingdom"
phylum_name = "phylum"
class_name = "class"
order_name = "order"
strain_name = "strain"

anonymous_read_id_name = "#anonymous_read_id"
sim_id_name = "sim_id"
read_name = "read"
read_id_name = "read_id"
prop_name = "proportion_genome_by_sim"

reads_file_name = "reads"
mapping_read_file_name = "mapping_read_genome"
reads_genomes_file_name = "reads_genomes"

# analyse read2genome
pred_name = "predict"
pred_detailed_name = "detailed_prediction"
pred_threshold_name = "prediction_threshold"
prop_pred_name = "proportion_pred"
prop_true_name = "proportion_true"
prob_name = "prob"
pair_name = "pair"
kmer_name = "kmer"

# metagenome2vec
id_fasta_name = "id.fasta"
id_subject_name = "id.subject"
group_name = "group"
count_name = "count"
genome_name = "genome"
control_category = "Control"
read_formatted_name = "read_formatted"

m2v_zip_name = "metagenome2vec.zip"

# NN
learning_rate = "learning_rate"
batch_size = "batch_size"
input_dim = "input_dim"
n_epoch = "n_epoch"
weight_decay = "weight_decay"
step_size = "step_size"
hidden_dim = "hidden_dim"
n_layer_before_flatten = "n_layer_before_flatten"
n_layer_after_flatten = "n_layer_after_flatten"
dropout = "dropout"
clip = "clip"
activation_function = "activation_function"
